### PR TITLE
Refactor exec steps (get, put, task)

### PR DESCRIPTION
**Pull request by pivotal-jwinters**
Piling onto our existing PR, we also removed the use of creds types in all the lower levels of the image/resource/exec packages. This way interpolation only ever happens in the exec steps.

### CODE REVIEW

File: atc/db/resource.go [hunk 1]
```diff
@@ -296,6 +297,35 @@ func (r *resource) SaveUncheckedVersion(version atc.Version, metadata ResourceCo
...
+	if err != nil {
+		return false, err
+	}
+
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = psql.Update("resource_config_versions").
```
Comment (vito) [inline atc/db/resource.go]: :thought_balloon: this seems a little iffy/imprecise but it's probably fine.

just thinking about the case where the pipeline is re-configured while the build is in-flight, with a different resource config, which happens to produce the same version identifiers, but runs with a different type or something such that the metadata is not accurate to the new resource type

the likelihood and impact of this kind of edge case is so low that it doesn't seem like something to worry about

Comment (vito) [reply atc/db/resource.go]: would matching on `version_md5` be faster here?

**Comment by cirocosta** [comment]
Hey @pivotal-jwinters ,

What's the rationale behind this PR? It seems like it contains some pretty significant changes, but I can't quickly tell from an outsider perspective 😬 

Thanks!!

**Comment by pivotal-jwinters** [comment]
@cirocosta I added some more info [here](https://github.com/concourse/concourse/issues/3788#issuecomment-503569022)

File: atc/db/resource.go (hunk 1 above)
Comment (pivotal-jwinters) [reply atc/db/resource.go]: Yeah good call on the `version_md5`. Updated.

File: atc/db/resource_config_factory.go [hunk 2]
```diff
@@ -109,22 +108,17 @@ func (f *resourceConfigFactory) FindOrCreateResourceConfig(
...
-	resourceTypes creds.VersionedResourceTypes,
+	resourceTypes atc.VersionedResourceTypes,
 ) (ResourceConfigDescriptor, error) {
 	resourceConfigDescriptor := ResourceConfigDescriptor{
 		Source: source,
 	}
 
 	customType, found := resourceTypes.Lookup(resourceTypeName)
 	if found {
-		source, err := customType.Source.Evaluate()
```
Comment (vito) [inline atc/db/resource_config_factory.go]: One concern I have here is that whereas before we would only evaluate the resource type that ends up being used, now we evaluate *all* of them.

Probably not a huge deal. Most resource types don't need credentials. :man_shrugging:

File: atc/db/team.go [hunk 3]
```diff
@@ -839,12 +839,17 @@ func (t *team) FindCheckContainers(logger lager.Logger, pipelineName string, res
 		return nil, nil, err
 	}
 
+	resourceTypes, err := creds.NewVersionedResourceTypes(variables, versionedResourceTypes).Evaluate()
```
Comment (vito) [inline atc/db/team.go]: (Yep this is what I was worried about, but again probably not a huge deal.)

File: atc/exec/get_step.go [hunk 4]
```diff
@@ -130,7 +130,10 @@ func (step *GetStep) Run(ctx context.Context, state RunState) error {
 		return err
 	}
 
-	resourceTypes := creds.NewVersionedResourceTypes(variables, step.plan.VersionedResourceTypes)
+	resourceTypes, err := creds.NewVersionedResourceTypes(variables, step.plan.VersionedResourceTypes).Evaluate()
```
Comment (vito) [inline atc/exec/get_step.go]: (Here too.)

File: atc/exec/put_step.go [hunk 5]
```diff
@@ -94,7 +94,10 @@ func (step *PutStep) Run(ctx context.Context, state RunState) error {
 		return err
 	}
 
-	resourceTypes := creds.NewVersionedResourceTypes(variables, step.plan.VersionedResourceTypes)
+	resourceTypes, err := creds.NewVersionedResourceTypes(variables, step.plan.VersionedResourceTypes).Evaluate()
```
Comment (vito) [inline atc/exec/put_step.go]: (...and here.)

File: atc/exec/task_step.go [hunk 6]
```diff
@@ -126,7 +125,11 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
 	})
 
 	variables := creds.NewVariables(step.secrets, step.metadata.TeamName, step.metadata.PipelineName)
-	resourceTypes := creds.NewVersionedResourceTypes(variables, step.plan.VersionedResourceTypes)
+
+	resourceTypes, err := creds.NewVersionedResourceTypes(variables, step.plan.VersionedResourceTypes).Evaluate()
```
Comment (vito) [inline atc/exec/task_step.go]: (...and here.)

File: atc/exec/task_step.go [hunk 7]
```diff
@@ -339,7 +342,7 @@ func (step *TaskStep) imageSpec(logger lager.Logger, repository *artifact.Reposi
 	} else if config.ImageResource != nil {
 		imageSpec.ImageResource = &worker.ImageResource{
 			Type:    config.ImageResource.Type,
-			Source:  creds.NewSource(boshtemplate.StaticVariables{}, config.ImageResource.Source),
```
Comment (vito) [inline atc/exec/task_step.go]: Ha, guessing this is 'cause it really just never supported interpolation here but the type demanded it. I guess this is a small win.

File: atc/radar/resource_scanner.go [hunk 8]
```diff
@@ -154,10 +154,15 @@ func (scanner *resourceScanner) scan(logger lager.Logger, resourceID int, fromVe
 		return 0, err
 	}
 
-	versionedResourceTypes := creds.NewVersionedResourceTypes(
+	versionedResourceTypes, err := creds.NewVersionedResourceTypes(
 		scanner.variables,
 		resourceTypes.Deserialize(),
-	)
+	).Evaluate()
```
Comment (vito) [inline atc/radar/resource_scanner.go]: (...and here.)

File: atc/radar/resource_type_scanner.go [hunk 9]
```diff
@@ -115,10 +115,15 @@ func (scanner *resourceTypeScanner) scan(logger lager.Logger, resourceTypeID int
 		return 0, err
 	}
 
-	versionedResourceTypes := creds.NewVersionedResourceTypes(
+	versionedResourceTypes, err := creds.NewVersionedResourceTypes(
 		scanner.variables,
 		resourceTypes.Deserialize(),
-	)
+	).Evaluate()
```
Comment (vito) [inline atc/radar/resource_type_scanner.go]: (...and here.)

File: atc/db/resource_config_factory.go (hunk 2 above)
Comment (vito) [reply atc/db/resource_config_factory.go]: I think the mitigation here is to just enable caching. That resource type is going to be actually evaluated and used *somewhere*, so it'll have been fetched anyway, and if you're using credential caching it'll only hit the credential manager once regardless of how many places call `.Evaluate` within that minute.

**Pull request merged**